### PR TITLE
feat: REST API

### PR DESCRIPTION
**Pull request by jawadst**
This PR adds support for a REST API with JSON:API standard.

It's in very early stage and still requires a lot of work to be complete:

- [x] Base REST support for CRUD
- [ ] Complete code coverage with missing tests
- [ ] Generate collection name from model (`user` -> `/users`) either via pluralize or by getting the collection name from the schema
- [ ] Ensure ts-japi is used for all serialization
- [ ] Compound IDs
- [ ] Filtering
- [ ] Pagination
- [ ] Links/relationships
- [ ] Generate Swagger

----

This PR introduces a new way to initialize APIs with Zenstack to allow support for more API standards. Example with Express:

```ts
import { ZenStackMiddleware } from '@zenstackhq/server/express';
import Prisma from '@zenstackhq/server/api/prisma';
import REST from '@zenstackhq/server/api/rest';

const { prisma, zodSchemas } = await loadSchema(schema);

const app = express();
app.use(bodyParser.json());

// Initialize Prisma-like API (renamed from "openapi")
app.use('/api', ZenStackMiddleware({
    getPrisma: () => prisma,
    zodSchemas,
    api: Prisma
}));

// Initialize REST API
app.use('/api', ZenStackMiddleware({
    getPrisma: () => prisma,
    zodSchemas,
    api: REST
}));
```

This PR also renames the API format "openapi" to "prisma" to better match what it actually does.

### CODE REVIEW

File: packages/server/src/api/rest/index.ts [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        };
+    }
+
+    method = method.toUpperCase();
+
+    try {
+        if (method === 'GET') {
+            // Handling GET requests
+            if (!resourceId) {
+                const items = (await prisma[model].findMany({ where: query })) as Item[];
```
Comment (ymc9) [inline packages/server/src/api/rest/index.ts:71]: `query` is actually the query parameters here, so not sure if passing it directly as `where` would work. The `Prisma` API handler used a `q` parameter to pass json-serialized prisma query. Do you it's good to do the same here?

**Review by ymc9 — COMMENTED** [review]
Thanks a lot for the contribution @jawadst ! Overall it looks very good! I've added a few comments; please check if they make sense to you.

For the other TODOs listed, I'm not sure how much time you have to continue working on them or if you think it's better for me to continue based on this foundation. Let me know.

File: packages/server/src/express/middleware.ts [hunk 2]
```diff
@@ -22,6 +22,11 @@ export interface MiddlewareOptions {
      * Zod schemas for validating request input. Pass `true` to load from standard location (need to enable `@core/zod` plugin in schema.zmodel).
      */
     zodSchemas?: ModelZodSchema | boolean;
+
+    /**
+     * API format to use from `@zenstackhq/server/api`
+     */
+    api: HandleRequestFn;
```
[the remark this answers is not in the document]
Comment (jawadst) [reply packages/server/src/express/middleware.ts]: Agreed and done in https://github.com/zenstackhq/zenstack/pull/326/commits/f8e97ef642976c784cbe407bbc052d856b28a4b0

File: packages/server/src/api/rest/index.ts (hunk 1 above)
Comment (jawadst) [reply packages/server/src/api/rest/index.ts:71]: The `query` there was kept from the original Prisma API code as I had not worked on this part yet.

JSON:API is not prescriptive on the topic. They do recommend using `filter[field]` in the query-string for filtering (https://jsonapi.org/recommendations/#filtering). That would mean that, by default, we would only support exact equality. Other types of filtering (>, <, etc.) would be on us to define.

Generally, REST APIs do not seem to have any standard on how the query-string parameters should be used for filtering/sorting/pagination and I have seen a little bit of everything. Here is an example of how Feathers does it: https://feathersjs.com/api/databases/querying.html

Given that there is no strong standard, I am inclined to do it as you describe (accept a JSON-encoded field that can be passed directly to prisma `findMany` function) as that will make our life much easier (we automatically get all of Prisma's features on the topic). While having a JSON-encoded field in the query-string is not very REST, I think it's worth it given how much it simplifies the implementation.

What do you think?

File: packages/server/src/api/rest/index.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    path,
+    query,
+    requestBody,
+    prisma,
+    logger,
+    zodSchemas,
+}: RequestContext): Promise<Response> {
+    const parts = path.split('/').filter((p) => !!p);
+    const model = parts[0];
+    const resourceId = parts[1];
```
[the remark this answers is not in the document]
Comment (jawadst) [reply packages/server/src/api/rest/index.ts:56]: I hadn't though of this point. `[1, 2]` seems problematic if the IDs are strings that can contain `[` in them (also an edge case) as we would not be able to distinguish between a compound ID and a regular ID. I'll do a few tests to see if that'd would work.

[the remark this answers is not in the document]
Comment (ymc9) [reply packages/server/src/api/rest/index.ts:56]: Maybe we can skip the square brackets and just use a comma-separated list, `/api/foo/1,2`. As long as the ID doesn't contain comma (I think we can make it a requirement), we should be fine. What do you think?

File: packages/server/src/api/rest/index.ts (hunk 1 above)
Comment (ymc9) [reply packages/server/src/api/rest/index.ts:71]: Yes, I completely agree with you. Using a json-encoded query parameter is both easy to implement and able to preserve Prisma's flexibility. In the future we can add support for JSON:API style filtering and use it as a shortcut and override.

File: packages/server/src/api/rest/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (potion-cellar) [reply packages/server/src/api/rest/index.ts:56]: > .

Of course, if the query params are being used for filtering etc. then this wouldn't be the best way to go about it, I suppose.

File: packages/server/src/api/rest/index.ts (hunk 1 above)
Comment (ymc9) [reply packages/server/src/api/rest/index.ts:71]: Hey @jawadst . I think the PR is in pretty good shape now. I'm going to merge it into the feature branch and continue making some tweaks there.

The main difference between the `Prisma` and `REST` is the usage of JSON:API standard. I'll probably merge these two implementations and add a switch to control the data protocol. In the future we can make the JSON:API flavor the default one.
